### ISSUE TITLE
`systemd-importd` fails to rename image after `machinectl pull-tar`

### ISSUE THREAD

**Issue by rezib** [systemd/systemd#21397]
**systemd version the issue has been seen with**

> ```
> # systemctl --version
> systemd 248 (248.3-1ubuntu8)
> +PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS -OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP -LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
> ```

**Used distribution**
 > Ubuntu 21.10

**Linux kernel version used** (`uname -a`)
 > `Linux carbon 5.13.0-20-generic #20-Ubuntu SMP Fri Oct 15 14:21:35 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > amd64

**Expected behaviour you didn't see**
 > `machinectl pull-tar <img>` successfully import the tarball.

**Unexpected behaviour you saw**
 > Error message `Failed to rename to final image name to /var/lib/machines/.tar-https:\x2f\x2fhub\x2enspawn\x2eorg\x2fstorage\x2fdebian\x2fbullseye\x2ftar\x2fimage\x2etar\x2exz.\x222d9c858-5d07d4127fef9\x22: Permission denied

**Steps to reproduce the problem**
 > For example, run `machinectl pull-tar https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz debian-bullseye-tar` (w/ `--verify=no` if hub.nspawn.org GPG key is not imported in local keyring).

Note: I use systemd 248.3 with a workaround for #19115 using an override to emulate 86204ae:

```
# cat /etc/systemd/system/systemd-importd.service.d/local.conf 
[Service]
CapabilityBoundingSet=CAP_LINUX_IMMUTABLE
```

```
$ SYSTEMD_LOG_LEVEL=debug machinectl pull-tar https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz debian-bullseye-tar
Bus n/a: changing state UNSET → OPENING
sd-bus: starting bus by connecting to /run/dbus/system_bus_socket...
Bus n/a: changing state OPENING → AUTHENTICATING
Successfully forked off '(polkit-agent)' as PID 181514.
Skipping PR_SET_MM, as we don't have privileges.
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=3 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.898 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.import1 path=/org/freedesktop/import1 interface=org.freedesktop.import1.Manager member=PullTar cookie=4 reply_cookie=0 signature=sssb error-name=n/a error-message=n/a
Got message type=method_return sender=:1.899 destination=:1.898 path=n/a interface=n/a member=n/a cookie=6 reply_cookie=4 signature=uo error-name=n/a error-message=n/a
Enqueued transfer job 1. Press C-c to continue download in background.
Got message type=signal sender=org.freedesktop.DBus destination=:1.898 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.898 path=n/a interface=n/a member=n/a cookie=3 reply_cookie=2 signature=n/a error-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.import1',path='/org/freedesktop/import1',interface='org.freedesktop.import1.Manager',member='TransferRemoved' successfully installed.
Got message type=method_return sender=org.freedesktop.DBus destination=:1.898 path=n/a interface=n/a member=n/a cookie=4 reply_cookie=3 signature=n/a error-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.import1',interface='org.freedesktop.import1.Transfer',member='LogMessage' successfully installed.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=7 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Pulling 'https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz', saving as 'debian-bullseye-tar'.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=8 reply_cookie=0 signature=us error-name=n/a error-message=n/a
HTTP request to https://hub.nspawn.org/storage/debian/bullseye/tar/image.nspawn failed with code 404.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=9 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Settings file could not be retrieved, proceeding without.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=10 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Downloading 566B for https://hub.nspawn.org/storage/debian/bullseye/tar/SHA256SUMS.gpg.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=11 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Download of https://hub.nspawn.org/storage/debian/bullseye/tar/SHA256SUMS.gpg complete.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=12 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Downloading 45.6M for https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=13 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Downloading 79B for https://hub.nspawn.org/storage/debian/bullseye/tar/SHA256SUMS.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=14 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Download of https://hub.nspawn.org/storage/debian/bullseye/tar/SHA256SUMS complete.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=15 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Got 1% of https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=16 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Got 28% of https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz. 4s left at 6.9M/s.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=17 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Got 61% of https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz. 1s left at 9.8M/s.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=18 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Got 97% of https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz. 83ms left at 11.5M/s.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=19 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Download of https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz complete.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=20 reply_cookie=0 signature=us error-name=n/a error-message=n/a
SHA256 checksum of https://hub.nspawn.org/storage/debian/bullseye/tar/image.tar.xz is valid.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=21 reply_cookie=0 signature=us error-name=n/a error-message=n/a
gpg: Signature faite le jeu. 11 nov. 2021 07:14:26 CET
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=22 reply_cookie=0 signature=us error-name=n/a error-message=n/a
gpg:                avec la clef RSA E605F310C46032EBD9EDDDA09F9B61E3E4EF5957
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=23 reply_cookie=0 signature=us error-name=n/a error-message=n/a
gpg: Bonne signature de « Nspawn.org Team (Nspawn.org Signing Key for containers.) <team@nspawn.org> » [inconnu]
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=24 reply_cookie=0 signature=us error-name=n/a error-message=n/a
gpg: Attention : utilisation d'une clef sans confiance.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=25 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Signature verification succeeded.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=26 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Failed to rename to final image name to /var/lib/machines/.tar-https:\x2f\x2fhub\x2enspawn\x2eorg\x2fstorage\x2fdebian\x2fbullseye\x2ftar\x2fimage\x2etar\x2exz.\x222d9c858-5d07d4127fef9\x22: Opération non permise
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1/transfer/_1 interface=org.freedesktop.import1.Transfer member=LogMessage cookie=27 reply_cookie=0 signature=us error-name=n/a error-message=n/a
Exiting.
Got message type=signal sender=:1.899 destination=n/a path=/org/freedesktop/import1 interface=org.freedesktop.import1.Manager member=TransferRemoved cookie=28 reply_cookie=0 signature=uos error-name=n/a error-message=n/a
Bus n/a: changing state RUNNING → CLOSED
```

(Sorry for the mixed french output, I tried setting `LANG` to `en` but had no effect)

It seems to be due to the immutable attribute introduced in d8f9686:

```
# lsattr -d /var/lib/machines/.#tarab33d6ad60a36576/
----i---------e------- /var/lib/machines/.#tarab33d6ad60a36576/
```

Trying to manually rename the directory with `mv` results in the same error on `renameat2()` syscall:

```
# mv /var/lib/machines/.#tarab33d6ad60a36576 /var/lib/machines/foo
mv: cannot move '/var/lib/machines/.#tarab33d6ad60a36576' to '/var/lib/machines/foo': Operation not permitted
```

```
# strace --trace=renameat2 mv /var/lib/machines/.#tarab33d6ad60a36576 /var/lib/machines/foo
renameat2(AT_FDCWD, "/var/lib/machines/.#tarab33d6ad60a36576", AT_FDCWD, "/var/lib/machines/foo", RENAME_NOREPLACE) = -1 EPERM (Operation not permitted)
mv: cannot move '/var/lib/machines/.#tarab33d6ad60a36576' to '/var/lib/machines/foo': Operation not permitted
+++ exited with 1 +++
```

Removing the immutable attribute makes `mv` happy:

```
# chattr -i /var/lib/machines/.#tarab33d6ad60a36576
# mv /var/lib/machines/.#tarab33d6ad60a36576 /var/lib/machines/foo && echo ok
ok
```

**Comment by poettering** [comment]
Fixed already as part of c40d82abf7b23803aa7394a7a7e24c40c32af851, i.e. the new install_file() gets this right.

**Comment by averyfreeman** [comment]
This issue persists on Ubuntu 22.04 on ZFS, which is not scheduled to be released until March 2022. Verified working with btrfs when mounted btrfs-formatted zvol to /var/lib/machines as workaround. Any plans to support ZFS?
